src/main/java/com/github/wovnio/wovnjava/PathUrlLanguagePatternHandler.java

```diff
@@ -6,64 +6,85 @@
-    private ArrayList<Lang> supportedLangs;
+    private LanguageAliases languageAliases;
     private String sitePrefixPath;
     private Pattern getLangPattern;
     private Pattern matchSitePrefixPathPattern;
 
-    PathUrlLanguagePatternHandler(Lang defaultLang, ArrayList<Lang> supportedLangs, String sitePrefixPath) {
+    PathUrlLanguagePatternHandler(Lang defaultLang, LanguageAliases languageAliases, String sitePrefixPath) {
         this.defaultLang = defaultLang;
-        this.supportedLangs = supportedLangs;
+        this.languageAliases = languageAliases;
         this.sitePrefixPath = sitePrefixPath;
         this.getLangPattern = this.buildGetLangPattern(sitePrefixPath);
         this.matchSitePrefixPathPattern = this.buildMatchSitePrefixPathPattern(sitePrefixPath);
     }
 
     Lang getLang(String url) {
-        Lang lang = this.getLangMatch(url, this.getLangPattern);
-        return (lang != null && this.supportedLangs.contains(lang)) ? lang : null;
+        if (!this.matchSitePrefixPathPattern.matcher(url).lookingAt()) {
+            return null;
+        }
+
+        String languageIdentifier = this.getLangMatch(url, this.getLangPattern);
+        Lang lang = this.languageAliases.getLang(languageIdentifier);
+        if (lang != null) {
+            return lang;
+        } else if (this.languageAliases.hasAliasForDefaultLang) {
+            return null;
+        } else {
+            return this.defaultLang;
+        }
     }
 
     String convertToDefaultLanguage(String url) {
         Lang currentLang = this.getLang(url);
         if (currentLang == null) {
             return url;
+        }
+
+        String newUrl = this.removeLang(url, currentLang);
+        if (this.languageAliases.hasAliasForDefaultLang) {
+            return this.insertLang(newUrl, this.defaultLang);
         } else {
-            return this.removeLang(url, currentLang.code);
+            return newUrl;
         }
     }
 
-    String convertToTargetLanguage(String url, Lang lang) {
-        Lang currentLang = this.getLangMatch(url, this.getLangPattern);
-        if (currentLang != null && this.supportedLangs.contains(currentLang)) {
-            url = this.removeLang(url, currentLang.code);
+    String convertToTargetLanguage(String url, Lang targetLang) {
+        if (targetLang == this.defaultLang) {
+            return this.convertToDefaultLanguage(url);
         }
-        return this.insertLang(url, lang.code);
-    }
 
-    private String removeLang(String url, String lang) {
-        if (lang.isEmpty()) return url;
+        String languageIdentifier = this.getLangMatch(url, this.getLangPattern);
+        Lang currentLang = this.languageAliases.getLang(languageIdentifier);
+        if (currentLang != null) {
+            String newUrl = this.removeLang(url, currentLang);
+            return this.insertLang(newUrl, targetLang);
+        } else if (this.languageAliases.hasAliasForDefaultLang) {
+            return url;
```

Comment:
What case is this handling? I'm wondering if it is reachable in practice.

Reply:
```
defaultLang: "ja"
langCodeAlias: "ja": "japan"
--------------------
convertToTargetLanguage("http://site.com/global/page", <English>)
line 58) languageIdentifier   -> "global"
line 59) currentLang          -> null
```
Because language alias for default language is set, the input URL "http://site.com/global/page" falls outside of the accepted URLs, and we cannot modify it. So we return the input URL untouched at line 63.

-----
The helper function `getLangMatch` changed, and it's name may not be fitting anymore. I'll see if I can make this part less confusing.

Reply:
I think I get it now. When the user sets a default lang alias, we assume they mean that the URLs to their source lang files MUST have the alias in it. So the language alias must always be resolvable by us.
`convertToTargetLanguage("http://site.com/global/page", <English>) => http://site.com/global/page`
without a default lang alias:
`convertToTargetLanguage("http://site.com/global/page", <English>) => http://site.com/en/global/page`

Reply:
> When the user sets a default lang alias, we assume they mean that the URLs to their source lang files MUST have the alias in it.

Correct. 

> So if the default language alias isn't found in the URL, it must be malformed or something and we should do nothing.

It results in the same behavior as for site prefix path and custom domain when the input URL is something that does not match what we're allowed to change.

I'll try to add some comments, and maybe organise the test cases better.

Reply:
I've added some comments to a few of the conditional branches that I think may be a bit confusing. I've also tried to name the tests more accurately.
I'm not sure it got a lot better, though.

Reply:
I guess the most important thing is there are tests that fail if we tried to change this behavior. We want to make clear that we have decided that language aliases mean the un-aliased version is no longer recognized.